### PR TITLE
Merge current scrape with previous results

### PR DESCRIPTION
**Pull request by livgust**
- add previous data if it was grabbed within a certain time frame
- add time stamps to each location

### CODE REVIEW

File: data/dataDefaulter.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+                    const lowerTimeBound =
+                        new Date() - secondsOfTolerance * 1000;
+                    if (
+                        cachedResult.timestamp &&
+                        cachedResult.timestamp >= lowerTimeBound
+                    ) {
+                        combinedResults.push(cachedResult);
+                    }
+                } else {
+                    combinedResults.push(cachedResult);
```
Comment (j256) [inline data/dataDefaulter.js:28]: So with 0 seconds of tolerance we always push?  Is that right?

File: data/dataDefaulter.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+        return combinedResults;
+    }
+}
+
+function generateKey(entry) {
+    let uniqueIdentifier = "";
+    ["name", "street", "city", "zip"].forEach((key) => {
+        if (entry[key]) {
+            uniqueIdentifier += `${entry[key]};`;
```
Comment (j256) [inline data/dataDefaulter.js:41]: How about adding some sort of separator like '|' so that we don't get overlaps however unlikely?

File: data/dataDefaulter.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
+function addDefaultsToResults(
```
Comment (j256) [inline data/dataDefaulter.js:1]: addOldDataToResults?

File: main.js [hunk 4]
```diff
@@ -7,120 +7,147 @@ const { addExtra } = require("puppeteer-extra");
...
+                    //print out the issue but don't fail, this way we still publish updates
+                    //for other locations even if this website's scrape doesn't work
+                    console.log(error);
+                    return null;
+                })
+            )
+        );
+        browser.close();
+        let scrapedResultsArray = [];
+        for (const result of results) {
```
Comment (j256) [inline main.js:60]: This doesn't care if results is null?  Just checking.

File: main.js [hunk 5]
```diff
@@ -7,120 +7,147 @@ const { addExtra } = require("puppeteer-extra");
...
-			});
-			return;
-		} else {
-			const params = {
-				Bucket: process.env.AWSS3BUCKETNAME,
-				Key: "data.json",
-				Body: JSON.stringify(responseJson),
-			};
+        let finalResultsArray = [];
+        if (process.argv.length <= 2) {
```
Comment (j256) [inline main.js:70]: This test seems somewhat arbitrary.  Anything else we can test on?  Maybe the number of scrapers in the list?

File: main.js [hunk 6]
```diff
@@ -7,120 +7,147 @@ const { addExtra } = require("puppeteer-extra");
...
+            //only add default data if we're not testing individual scrapers
+            finalResultsArray = dataDefaulter.addDefaultsToResults(
+                scrapedResultsArray,
+                cachedResults,
+                15 * 60 //15 minutes allowance for cached data
```
Comment (j256) [inline main.js:75]: Can we move this to the top as a constant?  I'm not sure what the right JS pattern is here.  Or make configurable?

File: main.js [hunk 7]
```diff
@@ -7,120 +7,147 @@ const { addExtra } = require("puppeteer-extra");
...
+            });
+            return;
+        } else {
+            const params = {
+                Bucket: process.env.AWSS3BUCKETNAME,
+                Key: "data.json",
+                Body: JSON.stringify(responseJson),
+            };
+
+            // Uploading files to the bucket
```
Comment (j256) [inline main.js:110]: Should we have sanity checks here?  Must have some data?

File: site-scrapers/MAImmunizations.js [hunk 8]
```diff
@@ -1,101 +1,119 @@
 const sites = require("../data/sites.json");
```
Comment (j256) [inline site-scrapers/MAImmunizations.js:1]: These changes should be in a different PR IMHO.

**Review by adpjay — APPROVED** [review]
Looks great. I ran it locally and got the expected results.
As far as the "secondsOfTolerance" field: What do you think about returning even those results and adding an indicator to the UI that the data is X minutes old and may be stale. The user can decide if the data is too old.
From the way it's run, it seems MAImmunizations.org would have a 4 hour nightly maintenance window ;)

File: data/dataDefaulter.js [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    secondsOfTolerance
+) {
+    if (!(cachedResults && cachedResults.length)) {
+        return scrapedResults;
+    } else {
+        const combinedResults = [];
+        const scrapeGrouping = {};
+        scrapedResults.forEach((result) => {
+            combinedResults.push(result);
+            scrapeGrouping[generateKey(result)] = 1;
```
Comment (johnhawkinson) [inline data/dataDefaulter.js:13]: I don't know what a `fooGrouping` is. Perhaps `resultFound` or `scrapedResultFound`?

File: main.js [hunk 10]
```diff
@@ -7,120 +7,146 @@ const { addExtra } = require("puppeteer-extra");
...
+              headless: chromium.headless,
+              ignoreHTTPSErrors: true,
+          });
 
-		const responseJson = {
-			results: finalResultsArray,
-		};
+    const gatherData = async () => {
+        const results = await Promise.all(
+            scrapers.map((scraper) =>
```
Comment (johnhawkinson) [inline main.js:49]: Perhaps we should be adding the timestamps here, rather than in each scraper.
Especially given the disparate maintenance.

File: main.js [hunk 11]
```diff
@@ -7,120 +7,146 @@ const { addExtra } = require("puppeteer-extra");
...
+        browser.close();
+        let scrapedResultsArray = [];
+        for (const result of results) {
+            if (Array.isArray(result)) {
+                scrapedResultsArray.push(...result);
+            } else if (result) {
+                //ignore nulls
+                scrapedResultsArray.push(result);
+            }
+        }
```
Comment (johnhawkinson) [inline main.js:67]: Why are we looping over the results here when we have an arrow function filtering at L50?
Seems like either this should move to the arrow function or we should loop over the scrapers and push each resul in the loop. Maybe there's a promise subtlety I'm missing.

File: site-scrapers/MAImmunizations.js [hunk 12]
```diff
@@ -1,413 +1,118 @@
...
+        console.log("Got the Mass. Heroku error page, giving up.");
+    }
+    if (pages.length < 1) {
+        console.log(
+            "No content matching our CSS selector (looking for nav.pagination)!"
+        );
+    }
+    const maxPage = await pages[pages.length - 1].evaluate(
+        (node) => node.innerText
+    );
```
Comment (johnhawkinson) [inline site-scrapers/MAImmunizations.js:30]: This is going to throw a lot of exceptions since the above error checks don't cause the function to fail.
Why can't we `return false` or, barring that, throw an explicit exception after L21 and L25?

File: site-scrapers/MAImmunizations.js [hunk 13]
```diff
@@ -1,413 +1,118 @@
...
+            await page.goto(
+                sites.MAImmunizations.website + "?page=" + pageNumber
+            );
+        }
+
+        const entries = await page.$$("div.justify-between.border-b");
+        for (const entry of entries) {
+            //each p has label: information
+            //TODO: add "special instructions" inside div.w-6/12
+            //UPDATE: ignoring "special instructions" since they all just say to get the same brand vaccine for both doses
```
Comment (johnhawkinson) [inline site-scrapers/MAImmunizations.js:46]: https://xkcd.com/2422/
*    Moderna before Pfizer, you'll be none the wiser. 
*    Pfizer before Moderna then you'll... rule ancient Smyrna.

File: site-scrapers/index.js [hunk 14]
```diff
@@ -2,24 +2,23 @@ const fs = require("fs");
...
     }
+} else {
+    const ls = fs
+        .readdirSync("./site-scrapers", { withFileTypes: true })
+        .filter((item) => !item.isDirectory())
+        .filter((item) => item.name !== "index.js")
+        .map((item) => item.name);
+
+    ls.map((fileName) => {
+        scraper = require(`./${fileName}`);
```
Comment (johnhawkinson) [inline site-scrapers/index.js:19]: `let scraper`

(Do we need a linter?)

File: test/dataDefaulter.js [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+        assert.deepStrictEqual(finalResults[2], realisticDefaultData[1]);
+    });
+});
+
+describe("Tolerance for stale data", () => {
+    const secondsOfTolerance = 60;
+    it("does not include stale data if specified", () => {
+        const timestampedDefault = {
+            ...realisticDefaultData[0],
+            timestamp: new Date() - (secondsOfTolerance + 1) * 1000,
```
Comment (johnhawkinson) [inline test/dataDefaulter.js:108]: Again, unspecified date arithmetic. 
I'm not sure what the `+ 1` is doing there.

**Review by johnhawkinson — COMMENTED** [review]
Minor nits.
Difficult to test without cherry-picking in `76433b` ;)

(It would have been much easier to review if the whitespace changes had been separate commits, or even a separate pull request.)

**Comment by livgust** [comment]
@johnhawkinson it's not helpful for you to modify the PR viewing settings to ignore whitespace changes?

@adpjay I'm nervous about sites that are added and then are removed permanently (or semi-permanently), like some on the Cape. Could we compromise and do a 1-day tolerance? I agree we should accompany this with an indicator on the front-end as well :)

Will check out the individual comments not at 3AM.

**Comment by johnhawkinson** [comment]
> @adpjay I'm nervous about sites that are added and then are removed permanently (or semi-permanently), like some on the Cape. Could we compromise and do a 1-day tolerance? I agree we should accompany this with an indicator on the front-end as well :)

It seems like there are a lot of reasons the front-end client should have access to older-than-1-day data.
It doesn't have to display it at all, or in the default case, but it seems totally reasonable to want the client to help a user answer the question, "When was the last time the Commonwealth had a site near me?"

If you do the cutoff here, that analysis is never going to be possible (well, I guess the client or some other tool could trawl all the historical data json files to build it, but ugh!), but if we toss that data in the client, it's easy to revisit the decision in whatever way.

And of course, implementation in the client is trivial.

Unless the concern is file size? I don't think we're there, but maybe?

(
> @johnhawkinson it's not helpful for you to modify the PR viewing settings to ignore whitespace changes?

Well, that and `git diff -b`, `-w`, `--word-diff=color`, &c. It's still not great, and doesn't help much when a single line gets split out into multiple. But this was just a parenthetical note. And opinions on coding standards vary, this is a project about expediency, &c., &c.)

File: data/dataDefaulter.js [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+        });
+
+        cachedResults.forEach((cachedResult) => {
+            if (!scrapeGrouping[generateKey(cachedResult)]) {
+                if (secondsOfTolerance) {
+                    const lowerTimeBound =
+                        new Date() - secondsOfTolerance * 1000;
```
[the remark this answers is not in the document]
Comment (livgust) [reply data/dataDefaulter.js:22]: Date() documentation lists subtraction as an example; leaving as-is.

File: data/dataDefaulter.js [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+
+function generateKey(entry) {
+    let uniqueIdentifier = "";
+    ["name", "street", "city", "zip"].forEach((key) => {
+        if (entry[key]) {
+            uniqueIdentifier += `${entry[key]};`;
+        }
+    });
+    //make lower case and replace ".", "-", "," and " " so we don't get TOO picky
+    uniqueIdentifier = uniqueIdentifier.toLowerCase().replace(/[\.,-\s]/g, "");
```
[the remark this answers is not in the document]
Comment (livgust) [reply data/dataDefaulter.js]: rationale was to not get multiple entries because someone said "rd" instead of "rd." I'll change to your \W suggestion.

File: site-scrapers/index.js (hunk 14 above)
Comment (livgust) [reply site-scrapers/index.js:19]: I have eslint running, and prettier. What's your question?

File: test/dataDefaulter.js (hunk 15 above)
Comment (livgust) [reply test/dataDefaulter.js:108]: making sure the timestamp is before the threshold (1 second older). I don't see anything saying subtraction isn't supported with Date().

**Review by livgust — COMMENTED** [review]
Updated PR addressing most comments

**Review by j256 — APPROVED** [review]

File: site-scrapers/index.js (hunk 14 above)
Comment (johnhawkinson) [reply site-scrapers/index.js:19]: I don't mean to belabor this, but `let scraper` on L6 is block scoped to L5-L10, so `scraper` on L19 is a global variable.
